src/Castle.Core/DynamicProxy/Generators/ClassProxyGenerator.cs

```diff
@@ -27,25 +27,21 @@ namespace Castle.DynamicProxy.Generators
 
 	internal class ClassProxyGenerator : BaseProxyGenerator
 	{
-		public ClassProxyGenerator(ModuleScope scope, Type targetType, ProxyGenerationOptions options)
-			: base(scope, targetType, options)
+		public ClassProxyGenerator(ModuleScope scope, Type targetType, Type[] interfaces, ProxyGenerationOptions options)
+			: base(scope, targetType, interfaces, options)
 		{
-			CheckNotGenericTypeDefinition(targetType, "targetType");
 			EnsureDoesNotImplementIProxyTargetAccessor(targetType, "targetType");
 		}
 
-		public Type GenerateCode(Type[] interfaces)
+		protected override CacheKey GetCacheKey()
 		{
-			interfaces = TypeUtil.GetAllInterfaces(interfaces);
-			CheckNotGenericTypeDefinitions(interfaces, "interfaces");
-			var cacheKey = new CacheKey(targetType, interfaces, ProxyGenerationOptions);
-			return ObtainProxyType(cacheKey, (n, s) => GenerateType(n, interfaces, s));
+			return new CacheKey(targetType, interfaces, ProxyGenerationOptions);
 		}
 
-		protected virtual Type GenerateType(string name, Type[] interfaces, INamingScope namingScope)
+		protected override Type GenerateType(string name, INamingScope namingScope)
```

Comment:
This would have made `ClassProxyGenerator` immutable if `ProxyGenerationOptions` was immutable.